src/middleware/immer.ts

```diff
@@ -35,25 +39,26 @@ type SkipTwo<T> = T extends { length: 0 }
 type WithImmer<S> = Write<S, StoreImmer<S>>
 
 type StoreImmer<S> = S extends {
-  getState: () => infer T
   setState: infer SetState
 }
-  ? SetState extends {
-      (...a: infer A1): infer Sr1
-      (...a: infer A2): infer Sr2
-    }
-    ? {
-        setState(
-          nextStateOrUpdater: T | Partial<T> | ((state: Draft<T>) => void),
-          shouldReplace?: false,
-          ...a: SkipTwo<A1>
-        ): Sr1
-        setState(
-          nextStateOrUpdater: T | ((state: Draft<T>) => void),
-          shouldReplace: true,
-          ...a: SkipTwo<A2>
-        ): Sr2
+  ? SetState extends StoreApi<infer T>['setState']
```

Comment:
I'm not sure if this works because `StoreApi['setState']` is very complicated type. Also, `setState` can be modified by upstream middleware. Can we infer `T` from `A1` and `A2`?

Reply:
It does infer correctly as the StoreApi<T> `setState` type is based off of the single generic passed into StoreApi. I think doing it this way bottles the complexity of the type and handles it appropriately. All the tests passed and it seems to resolve correctly. 

When I tried to base it off of A1 and A2, I would've needed to fight through the entire intersection type (e.g. `T | Partial<T> | ...`) in order to get T. That method feels more brittle than inferring through StoreApi.

Any situations you're thinking of specifically here? Like a modified `setState` type from another middleware?

Reply:
Well, actually, if I used the A2 type, there would be no Partial to fight with. Just the object and the function type. One sec, I'll try that.

Reply:
> Like a modified setState type from another middleware?

Yes.

Reply:
Check out my latest commit and let me know what you think of that. The assumption I made is that `A2` will always have a first parameter that is `T | (...args: any[]) => any`. So, `T` can be found by excluding any function types. That should be more robust.

Reply:
Inferring off of `A1` seemed non-trivial given that it is `T | Partial<T> | ...`, which I'm not sure if there's a method to derive `T` out of that.

Reply:
Hmmm, interesting. I guess the approach is right though.